### PR TITLE
HDFS-15374. Add documentation for command `fedbalance`.

### PR DESCRIPTION
**Pull request by wojiaodoubao**
HDFS-15374. Add documentation for command `fedbalance`.

### CODE REVIEW

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   0m 32s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| +0 :ok: |  markdownlint  |   0m  0s |  markdownlint was not available.  |
| +1 :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
||| _ trunk Compile Tests _ |
| +0 :ok: |  mvndep  |   1m 10s |  Maven dependency ordering for branch  |
| +1 :green_heart: |  mvninstall  |  20m  6s |  trunk passed  |
| +1 :green_heart: |  mvnsite  |   2m 36s |  trunk passed  |
| +1 :green_heart: |  shadedclient  |  37m 55s |  branch has no errors when building and testing our client artifacts.  |
||| _ Patch Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 26s |  Maven dependency ordering for patch  |
| +1 :green_heart: |  mvninstall  |   2m 25s |  the patch passed  |
| +1 :green_heart: |  mvnsite  |   2m 23s |  the patch passed  |
| -1 :x: |  whitespace  |   0m  0s |  The patch has 6 line(s) that end in whitespace. Use git apply --whitespace=fix <<patch_file>>. Refer https://git-scm.com/docs/git-apply  |
| +1 :green_heart: |  xml  |   0m  2s |  The patch has no ill-formed XML file.  |
| +1 :green_heart: |  shadedclient  |  13m 50s |  patch has no errors when building and testing our client artifacts.  |
||| _ Other Tests _ |
| +1 :green_heart: |  asflicense  |   0m 32s |  The patch does not generate ASF License warnings.  |
|  |   |  59m 48s |   |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.40 ServerAPI=1.40 base: https://builds.apache.org/job/hadoop-multibranch/job/PR-2035/1/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/2035 |
| Optional Tests | dupname asflicense mvnsite xml markdownlint |
| uname | Linux c9caa8420966 4.15.0-58-generic #64-Ubuntu SMP Tue Aug 6 11:12:41 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / c30c23cb665 |
| whitespace | https://builds.apache.org/job/hadoop-multibranch/job/PR-2035/1/artifact/out/whitespace-eol.txt |
| Max. process+thread count | 414 (vs. ulimit of 5500) |
| modules | C: hadoop-project hadoop-tools U: . |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-2035/1/console |
| versions | git=2.17.1 maven=3.6.0 |
| Powered by | Apache Yetus 0.12.0 https://yetus.apache.org |


This message was automatically generated.

File: hadoop-tools/hadoop-federation-balance/src/site/markdown/FederationBalance.md [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+
+  This will copy data from hdfs://nn0:8020/foo/src to hdfs://nn1:8020/foo/dst
+  incrementally and finally update the mount entry to:
+
+    /foo/src --> hdfs://nn1:8020/foo/dst
+
+  If the hadoop shell process exits unexpectedly and we want to continue the
+  unfinished job, we can use command:
+
+    bash$ /bin/hadoop fedbalance continue
```
Comment (goiri) [inline hadoop-tools/hadoop-federation-balance/src/site/markdown/FederationBalance.md:66]: Should this have code format like 73 (I think there might even be a better way).

File: hadoop-tools/hadoop-federation-balance/src/site/markdown/FederationBalance.md [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+ - [Architecture of Federation Balance](#Architecture_of_Federation_Balance)
+
+---
+
+Overview
+--------
+
+  Federation Balance is a tool balancing data across different federation
+  namespaces. It uses DistCp to copy data from the source path to the target
+  path. First it creates a snapshot at the source path and submit the initial
```
Comment (goiri) [inline hadoop-tools/hadoop-federation-balance/src/site/markdown/FederationBalance.md:34]: submits

File: hadoop-tools/hadoop-federation-balance/src/site/markdown/FederationBalance.md [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+ - [Configuration Options](#Configuration_Options)
+ - [Architecture of Federation Balance](#Architecture_of_Federation_Balance)
+
+---
+
+Overview
+--------
+
+  Federation Balance is a tool balancing data across different federation
+  namespaces. It uses DistCp to copy data from the source path to the target
```
Comment (goiri) [inline hadoop-tools/hadoop-federation-balance/src/site/markdown/FederationBalance.md:33]: link to distcp

File: hadoop-tools/hadoop-federation-balance/src/site/markdown/FederationBalance.md [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+  path. First it creates a snapshot at the source path and submit the initial
+  distcp. Then it uses distcp diff to do the incremental copy. Finally when the
+  source and the target are the same, it updates the mount table in Router and
+  move the source to trash.
+
+  This document aims to describe the design and usage of the Federation Balance.
```
Comment (goiri) [inline hadoop-tools/hadoop-federation-balance/src/site/markdown/FederationBalance.md:39]: In this doc, usage is first and then design.

File: hadoop-tools/hadoop-federation-balance/src/site/markdown/FederationBalance.md [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+  continue to execute them.
+  
+  If we want to balance in a normal federation cluster, use the command below.
+  
+    `bash$ /bin/hadoop fedbalance -router false submit hdfs://nn0:8020/foo/src hdfs://nn1:8020/foo/dst`
+    
+  The option `-router false` indicates this is not in router-based federation.
+  The source path must includes the source cluster.
+
+Command Options
```
Comment (goiri) [inline hadoop-tools/hadoop-federation-balance/src/site/markdown/FederationBalance.md:78]: This should be a subitem of usage right?

File: hadoop-tools/hadoop-federation-balance/src/site/markdown/FederationBalance.md [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+  See the License for the specific language governing permissions and
+  limitations under the License. See accompanying LICENSE file.
+-->
+
+Federation Balance Guide
+=====================
+
+---
+
+ - [Overview](#Overview)
```
Comment (goiri) [inline hadoop-tools/hadoop-federation-balance/src/site/markdown/FederationBalance.md:20]: Isn't toc already generating with some command?

File: hadoop-tools/hadoop-federation-balance/src/site/markdown/FederationBalance.md [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+### Balance Procedure Scheduler
+
+  The Balance Procedure Scheduler implements a state machine. It's responsible
+  for scheduling a balance job, including submit, run, delay and recover.
+  The model is showed below:
+
+  ![Balance Procedure Scheduler](images/BalanceProcedureScheduler.png)
+
+  * After a job is submitted, the job is added to the pendingQueue.
+  * Worker thread takes job and run it. Journals are written to storage.
```
Comment (goiri) [inline hadoop-tools/hadoop-federation-balance/src/site/markdown/FederationBalance.md:117]: The worker threads

File: hadoop-tools/hadoop-federation-balance/src/site/markdown/FederationBalance.md [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+  * If writing journal fails, the job is added to the recoverQueue for later
```
Comment (goiri) [inline hadoop-tools/hadoop-federation-balance/src/site/markdown/FederationBalance.md:118]: the journal

File: hadoop-tools/hadoop-federation-balance/src/site/markdown/FederationBalance.md [hunk 9]
```diff
@@ -0,0 +1,156 @@
...
+    recovery. If Worker thread catches a RetryTaskException, it adds the job to
+    the delayQueue.
+  * Rooster thread takes job from delayQueue and adds it back to pendingQueue.
+  * When a scheduler starts, it will scan all the unfinished jobs from
```
Comment (goiri) [inline hadoop-tools/hadoop-federation-balance/src/site/markdown/FederationBalance.md:122]: it scans

File: hadoop-tools/hadoop-federation-balance/src/site/markdown/FederationBalance.md [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+    journal and add them to the recoverQueue. The recover thread will recover
+    them from journal and add them back to pendingQueue.
```
Comment (goiri) [inline hadoop-tools/hadoop-federation-balance/src/site/markdown/FederationBalance.md:124]: the journal

File: hadoop-tools/hadoop-federation-balance/src/site/markdown/FederationBalance.md [hunk 11]
```diff
@@ -0,0 +1,156 @@
...
+### DistCpFedBalance
+
+  DistCpFedBalance is implemented as a job of the state machine. All the distcp
+  balance logic are implemented here. A DistCpFedBalance job consists of 3
+  procedures:
+
+  * DistCpProcedure: This is the first procedure. It handles all the data copy
+    works. There are 6 stages:
+    * PRE_CHECK: Do the pre-check of the src and dst path.
+    * Init Distcp: Create a snapshot of the source path and distcp it to the
```
Comment (goiri) [inline hadoop-tools/hadoop-federation-balance/src/site/markdown/FederationBalance.md:135]: Should this be two stages? One for snapshot and the other for distcp?

File: hadoop-tools/hadoop-federation-balance/src/site/markdown/FederationBalance.md [hunk 12]
```diff
@@ -0,0 +1,156 @@
...
+      target paths. If `-forceCloseOpen` is set, this stage will finish when
+      there is no diff between src and dst. Otherwise this stage only finishes
+      when there is no diff and no open files.  
+    * DISABLE_WRITE: Disable write operations so the src won't be changed. When
+      working in router mode, it is done by making the mount point readonly.
+      Otherwise then it is done by cancelling the `x` permission of the source
+      path.
+    * Final Distcp(optional): Force close all the open files and submit the
+      final distcp.
+    * FINISH: Cleanup works. If the 'x' permission is cancelled then restoring
```
Comment (goiri) [inline hadoop-tools/hadoop-federation-balance/src/site/markdown/FederationBalance.md:147]: Kind of weird that some stages are capitals and others not.

File: hadoop-tools/hadoop-federation-balance/src/site/markdown/FederationBalance.md [hunk 13]
```diff
@@ -0,0 +1,156 @@
...
+    * PRE_CHECK: Do the pre-check of the src and dst path.
+    * Init Distcp: Create a snapshot of the source path and distcp it to the
+      target.
+    * Diff Distcp: Submit distcp with `-diff` round by round to sync source and
+      target paths. If `-forceCloseOpen` is set, this stage will finish when
+      there is no diff between src and dst. Otherwise this stage only finishes
+      when there is no diff and no open files.  
+    * DISABLE_WRITE: Disable write operations so the src won't be changed. When
+      working in router mode, it is done by making the mount point readonly.
+      Otherwise then it is done by cancelling the `x` permission of the source
```
Comment (goiri) [inline hadoop-tools/hadoop-federation-balance/src/site/markdown/FederationBalance.md:143]: x permission?

File: hadoop-tools/hadoop-federation-balance/src/site/markdown/FederationBalance.md (hunk 6 above)
Comment (wojiaodoubao) [reply hadoop-tools/hadoop-federation-balance/src/site/markdown/FederationBalance.md:20]: Thanks your comments ! I did a test. If I removed the toc then there would be no toc in the generated html.

File: hadoop-tools/hadoop-federation-balance/src/site/markdown/FederationBalance.md (hunk 1 above)
Comment (wojiaodoubao) [reply hadoop-tools/hadoop-federation-balance/src/site/markdown/FederationBalance.md:66]: Sorry I didn't understand. Do you mean the documentation actually doesn't need the code format ? I tried to control the line within 81 characters except the command line and the tables. I'm sure if it's the right way. Hope for your further advice.

File: hadoop-tools/hadoop-federation-balance/src/site/markdown/FederationBalance.md (hunk 11 above)
Comment (wojiaodoubao) [reply hadoop-tools/hadoop-federation-balance/src/site/markdown/FederationBalance.md]: In the INIT_DISTCP stage it first creates snapshot then submits distcp. In the DIFF_DISTCP stage it also needs to create snapshot and submit distcp. So I combine the snapshot and distcp.

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   0m 31s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| +0 :ok: |  markdownlint  |   0m  0s |  markdownlint was not available.  |
| +1 :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
||| _ trunk Compile Tests _ |
| +0 :ok: |  mvndep  |   1m  5s |  Maven dependency ordering for branch  |
| +1 :green_heart: |  mvninstall  |  19m 14s |  trunk passed  |
| +1 :green_heart: |  mvnsite  |   2m 31s |  trunk passed  |
| +1 :green_heart: |  shadedclient  |  36m 45s |  branch has no errors when building and testing our client artifacts.  |
||| _ Patch Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 25s |  Maven dependency ordering for patch  |
| +1 :green_heart: |  mvninstall  |   2m 23s |  the patch passed  |
| +1 :green_heart: |  mvnsite  |   2m 17s |  the patch passed  |
| -1 :x: |  whitespace  |   0m  0s |  The patch has 14 line(s) that end in whitespace. Use git apply --whitespace=fix <<patch_file>>. Refer https://git-scm.com/docs/git-apply  |
| +1 :green_heart: |  xml  |   0m  1s |  The patch has no ill-formed XML file.  |
| +1 :green_heart: |  shadedclient  |  13m 33s |  patch has no errors when building and testing our client artifacts.  |
||| _ Other Tests _ |
| +1 :green_heart: |  asflicense  |   0m 32s |  The patch does not generate ASF License warnings.  |
|  |   |  58m 12s |   |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.40 ServerAPI=1.40 base: https://builds.apache.org/job/hadoop-multibranch/job/PR-2035/2/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/2035 |
| Optional Tests | dupname asflicense mvnsite xml markdownlint |
| uname | Linux f885cd2acc00 4.15.0-58-generic #64-Ubuntu SMP Tue Aug 6 11:12:41 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / b2200a33a6c |
| whitespace | https://builds.apache.org/job/hadoop-multibranch/job/PR-2035/2/artifact/out/whitespace-eol.txt |
| Max. process+thread count | 421 (vs. ulimit of 5500) |
| modules | C: hadoop-project hadoop-tools U: . |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-2035/2/console |
| versions | git=2.17.1 maven=3.6.0 |
| Powered by | Apache Yetus 0.12.0 https://yetus.apache.org |


This message was automatically generated.

File: hadoop-tools/hadoop-federation-balance/src/site/markdown/FederationBalance.md (hunk 6 above)
Comment (goiri) [reply hadoop-tools/hadoop-federation-balance/src/site/markdown/FederationBalance.md:20]: Take a look at:
https://hadoop.apache.org/docs/current/hadoop-project-dist/hadoop-hdfs/ShortCircuitLocalReads.html
It just has:

Comment (goiri) [reply hadoop-tools/hadoop-federation-balance/src/site/markdown/FederationBalance.md:20]: Check:
https://raw.githubusercontent.com/apache/hadoop/a55d6bba71c81c1c4e9d8cd11f55c78f10a548b0/hadoop-hdfs-project/hadoop-hdfs/src/site/markdown/ShortCircuitLocalReads.md

File: hadoop-tools/hadoop-federation-balance/src/site/markdown/FederationBalance.md (hunk 1 above)
Comment (goiri) [reply hadoop-tools/hadoop-federation-balance/src/site/markdown/FederationBalance.md:68]: My main concern is differentiating the commands themselves from the entries definition with different formats.
I guess the bash at the beginning gives that but there might be a proper way to define something like xml or text.

File: hadoop-tools/hadoop-federation-balance/src/site/markdown/FederationBalance.md [hunk 14]
```diff
@@ -0,0 +1,177 @@
...
+    bash$ /bin/hadoop fedbalance -router false submit hdfs://nn0:8020/foo/src hdfs://nn1:8020/foo/dst
+    
+  The option `-router false` indicates this is not in router-based federation.
+  The source path must includes the source cluster.
+
+### RBF Mode And Normal Federation Mode
+
+  The federation balance tool has 2 modes: 
+  
+  * the router-based federation mode(rbf mode).
```
Comment (goiri) [inline hadoop-tools/hadoop-federation-balance/src/site/markdown/FederationBalance.md:84]: Space before the parenthesis and probably use RBF in capitals.

File: hadoop-tools/hadoop-federation-balance/src/site/markdown/FederationBalance.md [hunk 15]
```diff
@@ -0,0 +1,177 @@
...
+| -map | Max number of concurrent maps to use for copy. |
+| -bandwidth | Specify bandwidth per map in MB. |
+| -moveToTrash | If `true` move the source path to trash after the job is done. Otherwise delete the source path directly. |
+
+### Configuration Options
+--------------------
+
+| Configuration key              | Description                          |
+| ------------------------------ | ------------------------------------ |
+| hadoop.hdfs.procedure.work.thread.num | The worker threads number of the BalanceProcedureScheduler. Default is `10`. |
```
Comment (goiri) [inline hadoop-tools/hadoop-federation-balance/src/site/markdown/FederationBalance.md:117]: Maybe add a column for default.

File: hadoop-tools/hadoop-federation-balance/src/site/markdown/FederationBalance.md [hunk 16]
```diff
@@ -0,0 +1,177 @@
...
+### Command Options
+
+Command `submit` has 5 options:
+
+| Option key                     | Description                          |
+| ------------------------------ | ------------------------------------ |
+| -router | This option specifies the mode of the command. `True` indicates the router-based federation mode. `False` indicates the normal federation mode. |
+| -forceCloseOpen | If `true`, the DIFF_DISTCP stage forces close all open files when there is no diff. Otherwise it waits until there is no open files. The default value is `false`. |
+| -map | Max number of concurrent maps to use for copy. |
+| -bandwidth | Specify bandwidth per map in MB. |
```
Comment (goiri) [inline hadoop-tools/hadoop-federation-balance/src/site/markdown/FederationBalance.md:109]: Default column?

File: hadoop-tools/hadoop-federation-balance/src/site/markdown/FederationBalance.md (hunk 1 above)
Comment (wojiaodoubao) [reply hadoop-tools/hadoop-federation-balance/src/site/markdown/FederationBalance.md:60]: Seems the markdown doesn't have a rich support on text and bash script etc. All of them are format as code. So I add a Header: Source Destinations for the mount table. Would this be clear to differentiating the commands and the mount table?

File: hadoop-tools/hadoop-federation-balance/src/site/markdown/FederationBalance.md (hunk 14 above)
Comment (wojiaodoubao) [reply hadoop-tools/hadoop-federation-balance/src/site/markdown/FederationBalance.md]: done

File: hadoop-tools/hadoop-federation-balance/src/site/markdown/FederationBalance.md (hunk 15 above)
Comment (wojiaodoubao) [reply hadoop-tools/hadoop-federation-balance/src/site/markdown/FederationBalance.md]: done

File: hadoop-tools/hadoop-federation-balance/src/site/markdown/FederationBalance.md (hunk 16 above)
Comment (wojiaodoubao) [reply hadoop-tools/hadoop-federation-balance/src/site/markdown/FederationBalance.md]: done

**Comment by wojiaodoubao** [comment]
Hi @goiri , very sorry for my late response. Since the fedbalance function was still changing I wanted to update the doc until the fedbalance was finished. Now the fedbalance is done so I update the doc.
Some changes:
1. The option `-router` and `-forceCloseOpen` are changed to without arg. 
2. A new option `-delay` is added. 
3. Other comments are also addressed.
4. Add a title in the mount table to differentiate the command and the mount point.

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   0m 31s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| +0 :ok: |  markdownlint  |   0m  0s |  markdownlint was not available.  |
| +1 :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
||| _ trunk Compile Tests _ |
| +0 :ok: |  mvndep  |   1m  3s |  Maven dependency ordering for branch  |
| +1 :green_heart: |  mvninstall  |  18m 55s |  trunk passed  |
| +1 :green_heart: |  mvnsite  |   0m 52s |  trunk passed  |
| +1 :green_heart: |  shadedclient  |  34m 43s |  branch has no errors when building and testing our client artifacts.  |
||| _ Patch Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 25s |  Maven dependency ordering for patch  |
| +1 :green_heart: |  mvninstall  |   0m 38s |  the patch passed  |
| +1 :green_heart: |  mvnsite  |   0m 36s |  the patch passed  |
| -1 :x: |  whitespace  |   0m  0s |  The patch has 11 line(s) that end in whitespace. Use git apply --whitespace=fix <<patch_file>>. Refer https://git-scm.com/docs/git-apply  |
| +1 :green_heart: |  xml  |   0m  1s |  The patch has no ill-formed XML file.  |
| +1 :green_heart: |  shadedclient  |  13m 33s |  patch has no errors when building and testing our client artifacts.  |
||| _ Other Tests _ |
| +1 :green_heart: |  asflicense  |   0m 31s |  The patch does not generate ASF License warnings.  |
|  |   |  52m 41s |   |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.40 ServerAPI=1.40 base: https://builds.apache.org/job/hadoop-multibranch/job/PR-2035/3/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/2035 |
| Optional Tests | dupname asflicense mvnsite xml markdownlint |
| uname | Linux 22891428a048 4.15.0-58-generic #64-Ubuntu SMP Tue Aug 6 11:12:41 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / 9cbd76cc775 |
| whitespace | https://builds.apache.org/job/hadoop-multibranch/job/PR-2035/3/artifact/out/whitespace-eol.txt |
| Max. process+thread count | 413 (vs. ulimit of 5500) |
| modules | C: hadoop-project hadoop-tools/hadoop-federation-balance U: . |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-2035/3/console |
| versions | git=2.17.1 maven=3.6.0 |
| Powered by | Apache Yetus 0.12.0 https://yetus.apache.org |


This message was automatically generated.

**Comment by wojiaodoubao** [comment]
Refer fedbalance-site.xml in the doc.

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |  21m 51s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| +0 :ok: |  markdownlint  |   0m  0s |  markdownlint was not available.  |
| +1 :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
||| _ trunk Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 59s |  Maven dependency ordering for branch  |
| +1 :green_heart: |  mvninstall  |  19m 12s |  trunk passed  |
| +1 :green_heart: |  mvnsite  |   0m 54s |  trunk passed  |
| +1 :green_heart: |  shadedclient  |  35m  3s |  branch has no errors when building and testing our client artifacts.  |
||| _ Patch Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 28s |  Maven dependency ordering for patch  |
| +1 :green_heart: |  mvninstall  |   0m 40s |  the patch passed  |
| +1 :green_heart: |  mvnsite  |   0m 37s |  the patch passed  |
| -1 :x: |  whitespace  |   0m  0s |  The patch has 11 line(s) that end in whitespace. Use git apply --whitespace=fix <<patch_file>>. Refer https://git-scm.com/docs/git-apply  |
| +1 :green_heart: |  xml  |   0m  2s |  The patch has no ill-formed XML file.  |
| +1 :green_heart: |  shadedclient  |  13m 55s |  patch has no errors when building and testing our client artifacts.  |
||| _ Other Tests _ |
| +1 :green_heart: |  asflicense  |   0m 31s |  The patch does not generate ASF License warnings.  |
|  |   |  74m 48s |   |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.40 ServerAPI=1.40 base: https://builds.apache.org/job/hadoop-multibranch/job/PR-2035/4/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/2035 |
| Optional Tests | dupname asflicense mvnsite xml markdownlint |
| uname | Linux 0e84689aea83 4.15.0-58-generic #64-Ubuntu SMP Tue Aug 6 11:12:41 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / 9821b94c946 |
| whitespace | https://builds.apache.org/job/hadoop-multibranch/job/PR-2035/4/artifact/out/whitespace-eol.txt |
| Max. process+thread count | 420 (vs. ulimit of 5500) |
| modules | C: hadoop-project hadoop-tools/hadoop-federation-balance U: . |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-2035/4/console |
| versions | git=2.17.1 maven=3.6.0 |
| Powered by | Apache Yetus 0.12.0 https://yetus.apache.org |


This message was automatically generated.

**Comment by hadoop-yetus** [comment]
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   0m 30s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| +0 :ok: |  markdownlint  |   0m  0s |  markdownlint was not available.  |
| +1 :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
||| _ trunk Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 26s |  Maven dependency ordering for branch  |
| +1 :green_heart: |  mvninstall  |  19m 13s |  trunk passed  |
| +1 :green_heart: |  mvnsite  |   0m 52s |  trunk passed  |
| +1 :green_heart: |  shadedclient  |  34m 31s |  branch has no errors when building and testing our client artifacts.  |
||| _ Patch Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 26s |  Maven dependency ordering for patch  |
| +1 :green_heart: |  mvninstall  |   0m 40s |  the patch passed  |
| +1 :green_heart: |  mvnsite  |   0m 37s |  the patch passed  |
| +1 :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| +1 :green_heart: |  xml  |   0m  1s |  The patch has no ill-formed XML file.  |
| +1 :green_heart: |  shadedclient  |  13m 46s |  patch has no errors when building and testing our client artifacts.  |
||| _ Other Tests _ |
| +1 :green_heart: |  asflicense  |   0m 31s |  The patch does not generate ASF License warnings.  |
|  |   |  52m 43s |   |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.40 ServerAPI=1.40 base: https://builds.apache.org/job/hadoop-multibranch/job/PR-2035/5/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/2035 |
| Optional Tests | dupname asflicense mvnsite xml markdownlint |
| uname | Linux 041c9178452d 4.15.0-60-generic #67-Ubuntu SMP Thu Aug 22 16:55:30 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / b2facc84a1b |
| Max. process+thread count | 422 (vs. ulimit of 5500) |
| modules | C: hadoop-project hadoop-tools/hadoop-federation-balance U: . |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-2035/5/console |
| versions | git=2.17.1 maven=3.6.0 |
| Powered by | Apache Yetus 0.12.0 https://yetus.apache.org |


This message was automatically generated.

**Comment by hadoop-yetus** [comment]
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   0m 30s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| +0 :ok: |  markdownlint  |   0m  0s |  markdownlint was not available.  |
| +1 :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
||| _ trunk Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 26s |  Maven dependency ordering for branch  |
| +1 :green_heart: |  mvninstall  |  18m 53s |  trunk passed  |
| +1 :green_heart: |  mvnsite  |   0m 53s |  trunk passed  |
| +1 :green_heart: |  shadedclient  |  33m 59s |  branch has no errors when building and testing our client artifacts.  |
||| _ Patch Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 26s |  Maven dependency ordering for patch  |
| +1 :green_heart: |  mvninstall  |   0m 39s |  the patch passed  |
| +1 :green_heart: |  mvnsite  |   0m 36s |  the patch passed  |
| +1 :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| +1 :green_heart: |  xml  |   0m  2s |  The patch has no ill-formed XML file.  |
| +1 :green_heart: |  shadedclient  |  13m 52s |  patch has no errors when building and testing our client artifacts.  |
||| _ Other Tests _ |
| +1 :green_heart: |  asflicense  |   0m 31s |  The patch does not generate ASF License warnings.  |
|  |   |  52m 17s |   |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.40 ServerAPI=1.40 base: https://builds.apache.org/job/hadoop-multibranch/job/PR-2035/6/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/2035 |
| Optional Tests | dupname asflicense mvnsite xml markdownlint |
| uname | Linux cf9bcbcc0783 4.15.0-60-generic #67-Ubuntu SMP Thu Aug 22 16:55:30 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / 201d734af39 |
| Max. process+thread count | 422 (vs. ulimit of 5500) |
| modules | C: hadoop-project hadoop-tools/hadoop-federation-balance U: . |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-2035/6/console |
| versions | git=2.17.1 maven=3.6.0 |
| Powered by | Apache Yetus 0.12.0 https://yetus.apache.org |


This message was automatically generated.

**Comment by wojiaodoubao** [comment]
The review is done in HDFS-15374, so let's close this PR.

**Pull request closed without merging**